.github/workflows/tests.yaml

```diff
@@ -0,0 +1,263 @@
+        with:
+          path: ./examples/tutorial/basic-3/target
+          key: cargo-${{ runner.os }}-${{ hashFiles('./examples/tutorial/basic-3/**/Cargo.toml') }}
+      - uses: actions/cache@v2
+        name: basic-4 cache
+        id: cache-basic-4
+        with:
+          path: ./examples/tutorial/basic-4/target
+          key: cargo-${{ runner.os }}-${{ hashFiles('./examples/tutorial/basic-4/**/Cargo.toml') }}
+      - run: pushd examples/tutorial && yarn workspaces run test
+
+  setup-client-example:
+    needs: setup-anchor-cli
+    name: Setup Client Example Test
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        node:
+          - path: tests/events/
+            name: events.so
+          - path: examples/tutorial/basic-4/
+            name: basic_4.so
+          - path: examples/tutorial/basic-2/
+            name: basic_2.so
+          - path: tests/composite/
+            name: composite.so
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ./.github/actions/setup/
+      - uses: ./.github/actions/setup-solana/
+
+      - uses: actions/download-artifact@v2
+        with:
+          name: anchor-binary
+          path: ~/.cargo/bin/
+      - run: chmod +rwx ~/.cargo/bin/anchor
+      - run: pushd ${{ matrix.node.path }} && anchor build && popd
+      - uses: actions/upload-artifact@v2
+        with:
+          name: ${{ matrix.node.name }}
+          path: ${{ matrix.node.path }}target/deploy/${{ matrix.node.name }}
+  
+  test-client-example:
+    needs: setup-client-example
+    name: Client Example Test
+    runs-on: ubuntu-18.04
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ./.github/actions/setup/
+      - uses: ./.github/actions/setup-ts/
+
+      - uses: actions/download-artifact@v2
+        with:
+          name: anchor-binary
+          path: ~/.cargo/bin/
+      - uses: actions/download-artifact@v2
+        with:
+          name: events.so
+          path: tests/events/target/deploy/
+      - uses: actions/download-artifact@v2
+        with:
+          name: basic_4.so
+          path: examples/tutorial/basic-4/target/deploy/
+      - uses: actions/download-artifact@v2
+        with:
+          name: basic_2.so
+          path: examples/tutorial/basic-2/target/deploy/
+      - uses: actions/download-artifact@v2
+        with:
+          name: composite.so
+          path: tests/composite/target/deploy/
+      - uses: actions/cache@v2
+        name: Cache client/example target
+        id: cache-test-target
+        with:
+          path: client/example/target
+          key: cargo-${{ runner.os }}-client/example-${{ env.ANCHOR_VERSION }}
+      - uses: ./.github/actions/setup-solana/
+      - run: pushd client/example && ./run-test.sh && popd
+
+  test-programs:
+    needs: setup-anchor-cli
+    name: Test ${{ matrix.node.path }}
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        node:
+          - cmd: pushd tests/sysvars && anchor test && popd
```

Comment:
Do we really need `pushd` / `popd`? Every `cmd` should ignore previous changes in PATH :thinking: